File: ios/MapPolylineManager.swift
import GoogleMaps

private final class PolylineState {
  var current: RNPolyline
  var polyline: GMSPolyline?

  init(current: RNPolyline) {
    self.current = current
  }
}

final class MapPolylineManager {
  private let builder: MapPolylineBuilder
  private weak var mapView: GMSMapView?
  private var states: [String: PolylineState] = [:]
  private var destroyed = false

  init(builder: MapPolylineBuilder) {
    self.builder = builder
  }

  func attachMap(_ mapView: GMSMapView) {
    onMain {
      guard !self.destroyed else { return }
      self.mapView = mapView
      self.states.values.filter { $0.polyline == nil }.forEach { self.addToMap($0) }
    }
  }

  func add(_ polyline: RNPolyline) {
    onMain {
      guard !self.destroyed else { return }
      self.states.removeValue(forKey: polyline.id).map { self.removeFromMap($0) }
      let state = PolylineState(current: polyline)
      self.states[polyline.id] = state
      if self.mapView != nil {
        self.addToMap(state)
      }
    }
  }

  func update(_ next: RNPolyline) {
    onMain {
      guard !self.destroyed else { return }
      guard let state = self.states[next.id] else { return }
      let prev = state.current
      if prev.polylineEquals(next) { return }
      state.current = next
      state.polyline.map { self.builder.update(prev, next, $0) }
    }
  }

  func remove(id: String) {
    onMain {
      self.states.removeValue(forKey: id).map { self.removeFromMap($0) }
    }
  }

  func destroy() {
    onMain {
      self.destroyed = true
      self.states.values.forEach { self.removeFromMap($0) }
      self.states.removeAll()
      self.mapView = nil
    }
  }

  private func addToMap(_ state: PolylineState) {
    guard let mapView = self.mapView else { return }
    let polyline = self.builder.build(state.current)
    polyline.tagData = PolylineTag(id: state.current.id)
    polyline.map = mapView
    state.polyline = polyline
  }

  private func removeFromMap(_ state: PolylineState) {
    state.polyline?.map = nil
  }
}

File: android/src/main/java/com/rngooglemapsplus/MapPolylineManager.kt
package com.rngooglemapsplus

import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.Polyline
import com.rngooglemapsplus.extensions.polylineEquals

private class PolylineState(
  var current: RNPolyline,
) {
  var polyline: Polyline? = null
}

class MapPolylineManager(
  private val builder: MapPolylineBuilder,
) {
  private var map: GoogleMap? = null
  private val states = mutableMapOf<String, PolylineState>()
  private var destroyed = false

  fun attachMap(map: GoogleMap) =
    onUi {
      if (destroyed) return@onUi
      this.map = map
      states.values.filter { it.polyline == null }.forEach { addToMap(it) }
    }

  fun add(polyline: RNPolyline) =
    onUi {
      if (destroyed) return@onUi
      remove(polyline.id)
      val state = PolylineState(polyline)
      states[polyline.id] = state
      if (map != null) addToMap(state)
    }

  fun update(next: RNPolyline) =
    onUi {
      if (destroyed) return@onUi
      val state = states[next.id] ?: return@onUi
      val prev = state.current
      if (prev.polylineEquals(next)) return@onUi
      state.current = next
      state.polyline?.let { builder.update(prev, next, it) }
    }

  fun remove(id: String) =
    onUi {
      states.remove(id)?.let { removeFromMap(it) }
    }

  fun destroy() =
    onUi {
      destroyed = true
      states.values.forEach { removeFromMap(it) }
      states.clear()
      map = null
    }

  private fun addToMap(state: PolylineState) {
    state.polyline =
      map?.addPolyline(builder.build(state.current))?.apply {
        tag = PolylineTag(id = state.current.id)
      }
  }

  private fun removeFromMap(state: PolylineState) {
    state.polyline?.remove()
  }
}
